### ISSUE TITLE
Massive CPU/Memory/Disk usage with Ubuntu definitions update

### ISSUE THREAD

**Issue by ghost** [quay/clair#548]
**Description of Problem**:
Since the 26th, I've noticed Clair getting out of control with the update process - It continually utilizes the disk with reads until killed. Our Clair instance is using an AWS EBS volume (1000/3000 IOPS), and the reads continue at max capacity until the whole instance is starved. Most of our 4GB of memory also ends up getting used. The process goes haywire during an Ubuntu bzr fetch (a bummer as we don't use Ubuntu-based images anyway). The AWS instance using this is a c4.large.

![ebs_read_throughput](https://user-images.githubusercontent.com/37420850/39533833-4fc73e58-4ded-11e8-8e2d-610a684df61b.png)

I'm presently using Clair/Postgres with a simple docker-compose.yml file [1].

I will attach log files the next time this occurs - I accidentally removed the container before saving the logs.

As this started occurring on the 26/27th, perhaps the Ubuntu 18.04 release has something to do with it?

This appears to be related to #415 and #437.

**Environment**:
- Clair version/image: quay.io/coreos/clair:latest 
- Clair client name/version: 
- Host OS: Amazon Linux 1
- Kernel (e.g. `uname -a`): Linux ip-[snip] 4.14.33-51.34.amzn1.x86_64 #1 SMP Fri Apr 13 18:18:26 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
- Kubernetes version (use `kubectl version`): N/A, using docker-compose
- Helm version (use `helm version`): N/A
- Network/Firewall setup: AWS Security Groups for inbound connections, no blocking for outbound.

[1] [config.yaml.txt](https://github.com/coreos/clair/files/1968009/config.yaml.txt)

**Comment by dustyhorizon** [comment]
I am experiencing this too where bzr constantly hitting ~98% cpu usage.

I don't seem to see ubuntu completing its update when bzr is soaking up the resources.

**Comment by jbergstroem** [comment]
I see the same thing on a machine I've deployed on azure (4 cores hence 25%) using clair 2.0.2:
<img width="373" alt="screen shot 2018-05-11 at 10 38 35 am" src="https://user-images.githubusercontent.com/176984/39927080-8d2408e6-5507-11e8-9050-84f400b56e43.png">

ps output:
`clair     126248 99.2  2.9 310332 243100 ?       R    13:36   1:00 /usr/bin/python /usr/bin/bzr log --verbose -r1.. -n0`

**Comment by jbergstroem** [comment]
Docker logs aren't very interesting other than the fact that ubuntu takes hours to complete (output slightly snipped and ran through jq) where others are in the second range:
```json
[
  {
    "Event": "updating vulnerabilities",
    "Level": "info",
    "Location": "updater.go:182",
    "Time": "2018-05-14 00:37:21.005864"
  },
  {
    "Event": "fetching vulnerability updates",
    "Level": "info",
    "Location": "updater.go:228",
    "Time": "2018-05-14 00:37:21.005919"
  },
  {
    "Event": "Start fetching vulnerabilities",
    "Level": "info",
    "Location": "rhel.go:92",
    "Time": "2018-05-14 00:37:21.006092",
    "package": "RHEL"
  },
  {
    "Event": "Start fetching vulnerabilities",
    "Level": "info",
    "Location": "ubuntu.go:88",
    "Time": "2018-05-14 00:37:21.006070",
    "package": "Ubuntu"
  },
  {
    "Event": "Start fetching vulnerabilities",
    "Level": "info",
    "Location": "alpine.go:52",
    "Time": "2018-05-14 00:37:21.006191",
    "package": "Alpine"
  },
  {
    "Event": "Start fetching vulnerabilities",
    "Level": "info",
    "Location": "debian.go:63",
    "Time": "2018-05-14 00:37:21.006292",
    "package": "Debian"
  },
  {
    "Event": "Start fetching vulnerabilities",
    "Level": "info",
    "Location": "oracle.go:119",
    "Time": "2018-05-14 00:37:21.006003",
    "package": "Oracle Linux"
  },
  {
    "Event": "finished fetching",
    "Level": "info",
    "Location": "updater.go:242",
    "Time": "2018-05-14 00:37:21.646908",
    "updater name": "rhel"
  },
  {
    "Event": "finished fetching",
    "Level": "info",
    "Location": "updater.go:242",
    "Time": "2018-05-14 00:37:21.733195",
    "updater name": "alpine"
  },
  {
    "Event": "finished fetching",
    "Level": "info",
    "Location": "updater.go:242",
    "Time": "2018-05-14 00:37:22.106048",
    "updater name": "debian"
  },
  {
    "Event": "finished fetching",
    "Level": "info",
    "Location": "updater.go:242",
    "Time": "2018-05-14 00:37:22.438692",
    "updater name": "oracle"
  },
  {
    "Event": "finished fetching",
    "Level": "info",
    "Location": "updater.go:242",
    "Time": "2018-05-14 03:58:04.287160",
    "updater name": "ubuntu"
  },
  {
    "Event": "adding metadata to vulnerabilities",
    "Level": "info",
    "Location": "updater.go:268",
    "Time": "2018-05-14 03:58:04.551095"
  },
  {
    "Event": "fetcher note",
    "Level": "warning",
    "Location": "updater.go:204",
    "Time": "2018-05-14 04:09:52.803084",
    "note": "Ubuntu bionic is not mapped to any version number (eg. trusty->14.04). Please update me."
  },
  {
    "Event": "update finished",
    "Level": "info",
    "Location": "updater.go:213",
    "Time": "2018-05-14 04:09:52.806411"
  }
]
```

**Comment by yebinama** [comment]
Wasn't Ubuntu Bionic released on April 26th?
Maybe it's a coincidence but this warning refers to it:

{
    "Event": "fetcher note",
    "Level": "warning",
    "Location": "updater.go:204",
    "Time": "2018-05-14 04:09:52.803084",
    "note": "Ubuntu bionic is not mapped to any version number (eg. trusty->14.04). Please update me."
  }

**Comment by jbergstroem** [comment]
> @yebinama said: Wasn't Ubuntu Bionic released on April 26th?

Correct, corresponding PR is #553.

**Comment by dustyhorizon** [comment]
Tried clair-git:master docker image but issue persisted

**Comment by jbergstroem** [comment]
> @dustyhorizon said: Tried clair-git:master docker image but issue persisted

I don't think this is related to Bionic. My guess would be an issue working with the bzr repo from launchpad. I will `strace` the next stall and see what it says.

**Comment by jbergstroem** [comment]
So, looks like it's tied to parsing `bzr log` output (as `ps aux` shows). Can [parsing the output](https://github.com/coreos/clair/blob/6663bcef27b0a4221831a1ff7a05477247188b09/ext/vulnsrc/ubuntu/ubuntu.go#L271..L289) really be that slow? We're talking 14752 commits.

I logged into the container to make sure that the repo was up to date (`bzr pull`, ..) - which it was.

**Comment by jbergstroem** [comment]
@dustyhorizon, @yebinama are you populating your databases from scratch?

**Comment by jzelinskie** [comment]
I know launchpad is slow; perhaps the update is timing out and we're immediately retrying. bzr is probably the worst SCM that I've ever encountered, so I wouldn't be surprised if it does a bunch of disk operations before/during pulling down remote state.

**Comment by jbergstroem** [comment]
> @jzelinskie said: I know launchpad is slow, perhaps the update is timing out and we're immediately retrying. 

This looks like post network retrieval to me. I see no network activity during `bzr log` (or its so low `docker stats` stops being useful). I don't know `bzr` well enough to say if `bzr log --verbose -r1.. -n0` talks to wire, but in-container `netstat` didn't show much activity either.

Attaching a 5min ish strace log for the curious.
[trace.log.gz](https://github.com/coreos/clair/files/2002260/trace.log.gz)

Edit: confirmed near-zero activity by inspecting `/proc/$pid/net/netstat`

Edit2: I ran
```console
# time /usr/bin/bzr log --verbose -r1.. -n0 > /tmp/output
^Creal	0m 59.27s
user	0m 57.59s
sys	0m 1.66s
```
..to see if it was consuming output or generating output that was slow. The resulting file was 10Kb.

Writing to disk from within the container doesn't seem slow:
```console

# time dd if=/dev/urandom of=/tmp/out bs=4M count=1
1+0 records in
1+0 records out
real	0m 0.02s
user	0m 0.00s
sys	0m 0.02s
```

So, is this about reading the `bzr` repo contents? If if truly is this slow, perhaps its better to batch-read revisions in iterations of "N" so we can partially commit it to the database and continue from a newer revision. This way we won't start from scratch at the next attempt (when clair chooses to time it out).

**Comment by jzelinskie** [comment]
It could be related to [this](https://bugs.launchpad.net/bzr/+bug/367545). It looks like bzr eats huge amounts of memory and the iops could be your system swapping to disk.

**Comment by jbergstroem** [comment]
> @jzelinskie said: It could be related to [this](https://bugs.launchpad.net/bzr/+bug/367545).

Perhaps, but that seems to address `checkout`/`branch`. Are you saying that `log` would invoke it?

> @jzelinskie said: It looks like bzr eats huge amounts of memory and the iops could be your system swapping to disk.

At least on my system, there is very little disk activity and no swapping is occurring.

**Comment by ghost** [comment]
According to this launchpad report[1], Clair should be able to use git to clone:


> 
> Hi, thanks for you interest in the ubuntu-cve-tracker. The tracker has been converted to git and is available from https://code.launchpad.net/~ubuntu-security/ubuntu-cve-tracker/+git/ubuntu-cve-tracker .
> 
> I'm not quite sure why the high CPU load was seen during the bzr clone stage; however, during the conversion I did discover the cve tracker bazaar branch on launchpad was in an old v1 format which may have been part of the cause.
> 
> The layout/format of the tree has not changed. Please report bugs if yo find anything weird with the conversion. Thanks again!


[1] https://bugs.launchpad.net/ubuntu-cve-tracker/+bug/1771169

**Comment by jzelinskie** [comment]
Fixed in #568. Please open a new issue if you're still experiencing this bug on a non-bzr Clair.
